### PR TITLE
Script to import Users from CSV

### PR DESCRIPTION
**Pull request by NikhilM98**
Run `npm run seed:users-csv -- filename.csv` to seed data from CSV.

- [x] Import and validate CSV
- [x] Seed data in the database
- [x] Generate CSV with seeded user data
- [x] Finish randomization and validation

### CODE REVIEW

**Comment by NikhilM98** [comment]
Test:
Screenshot from 2019-04-12 23-38-40

Test data:
Screenshot from 2019-04-12 23-40-31

File: scripts/seed_users.js [hunk 1]
```diff
@@ -0,0 +1,414 @@
...
+}
+
+// Validate color string
+function isValidColor(color) {
+    try {
+        color = JSON.parse(color);
+    } catch (e) {
+        return false;
+    }
+    if (!color.stroke || !color.fill) return false;
```
Comment (llaske) [inline scripts/seed_users.js]: Hmmm. Ideally stroke and fill should have value in the list of valid color set here;https://github.com/llaske/sugarizer-server/blob/dev/dashboard/helper/xocolors.js#L3

File: scripts/seed_users.js [hunk 2]
```diff
@@ -0,0 +1,414 @@
...
+    return true
+}
+
+// Generate random password
+function generatePassword(length) {
+    var password = "";
+    var possible = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789";
```
Comment (llaske) [inline scripts/seed_users.js]: Not all alphanumeric character is allowed, see https://github.com/llaske/sugarizer-server/pull/117

File: scripts/seed_users.js [hunk 3]
```diff
@@ -0,0 +1,414 @@
...
+
+    // ToDo: Validate language from an array of valid languages
+    user.language = cleanString(user.language);
+    if (!user.language) {
+        user.language = "en";
+    }
+
+    user.color = cleanString(user.color);
+    if (!isValidColor(user.color)) {
+        // ToDo: Randomize Color
```
Comment (llaske) [inline scripts/seed_users.js]: Randomize will be just to get a value in the valid color array mentioned above

File: scripts/seed_users.js [hunk 4]
```diff
@@ -0,0 +1,414 @@
...
+        user.color = '{ "stroke":"#ff0000", "fill":"#0000ff" }';
+    }
+
+    // ToDo: Get minimum password length from .ini file
+    if (!user.password || typeof user.password != "string" || user.password.length < 4) {
```
Comment (llaske) [inline scripts/seed_users.js]: Password size should come from settings https://github.com/llaske/sugarizer-server/tree/dev#server-settings

File: scripts/seed_users.js [hunk 5]
```diff
@@ -0,0 +1,414 @@
...
+    return '{"name":"' + user.name + '", ' + 
+            '"color":' + user.color + ', ' +
+            '"role":"' + user.type + '", ' +
+            '"password":"' + user.password + '", ' + 
+            '"language":"' + user.language + '", ' +
+            '"options":{"sync":true, "stats":true}}';
+}
+
+// Initiate Master Admin
+var masterAdmin = stringifyUser(new User('Master_Admin_' + (+new Date()).toString(), 'admin', 'en', '{"stroke":"#ff0000", "fill":"#0000ff"}', 'password'));
```
Comment (llaske) [inline scripts/seed_users.js]: Does it mean that the script will create a master admin each time it's run?

File: scripts/seed_users.js [hunk 6]
```diff
@@ -0,0 +1,414 @@
...
+function stringifyClassroom(name) {
+    var newStudents = Classrooms[name].students;
+    var oldStudents = [];
+    if (Classrooms[name].data && Classrooms[name].data.students && Classrooms[name].data.students.length > 0) {
+        oldStudents = Classrooms[name].data.students 
+    }
+    var union = [...new Set([...newStudents, ...oldStudents])];
+    var students = JSON.stringify(union);
+    console.log(union);
+    return '{"name":"' + name + '","color":{"stroke":"#FF0000","fill":"#0000FF"},"students":' + students + '}';
```
Comment (llaske) [inline scripts/seed_users.js]: Will be good to randomize classroom color too.

File: scripts/seed_users.js (hunk 5 above)
Comment (NikhilM98) [reply scripts/seed_users.js]: Yes. The master admin will be deleted when the script finishes.
There's no way to access other API routes without an admin role. I prefer not to modify the existing API.

**Comment by NikhilM98** [comment]
Remaining Tasks:
- Randomize colors
- Get password min-length from .ini
- Validate Language
- Validate Colors

Current Test Screenshots:
- Terminal:
Screenshot from 2019-04-13 14-24-31

- Input Data:
Screenshot from 2019-04-13 14-22-05

- Output Data:
Screenshot from 2019-04-13 14-21-51

- Dashboard/Users (Before Inserting):
Screenshot from 2019-04-13 14-19-57

- Dashboard/Users (After Inserting) -- New users added:
Screenshot from 2019-04-13 14-28-34


- Dashboard/Classroom (Before Inserting):
Screenshot from 2019-04-13 14-20-05

- Dashboard/Classroom (After Inserting) -- Check update data and users count :
Screenshot from 2019-04-13 14-30-12

**Comment by NikhilM98** [comment]
@llaske Please review the PR.

Randomization in colors:
Screenshot from 2019-04-13 17-05-47
Screenshot from 2019-04-13 17-08-06

Terminal output:
Screenshot from 2019-04-13 17-09-16

Input CSV (Screenshot):
Screenshot from 2019-04-13 17-16-12


Output CSV (Screenshot):
Screenshot from 2019-04-13 17-16-37

Input CSV (RAW):
name,type,language,color,password,classroom
Nikhil1,admin,fr,Red,Pass001,asas
Nikhil2,student,ge,Green,Pass002,
,student,en,"{""stroke"":""#FF0000"",""fill"":""#0000FF""}",Pass003,
Nikhil1,student,es,Red,Pass004,asas
Nikhil5,student,French,"{""stroke"":""#FF0000"",""fill"":""#0000FF""}",Pass005,
Nikhil6,student,en,Blue,006, Some Class
Nikhil7,student,DE,"{""stroke"":""#00A0FF"",""fill"":""#FF2B34""}",Pass007,asas
Nikhil8,student,en,Green,Pass008,MyClassroom
Nikhil9,admin,de,Blue,Pass009,MyClassroom
Nikhil10,student,PT,Red,Pas,
Nikhil11,student,en,Green,Pass011,
Nikhil12,student,en,Blue,Pass012,Some Class

File: scripts/seed_users.js (hunk 2 above)
Comment (NikhilM98) [reply scripts/seed_users.js]: Done

File: scripts/seed_users.js (hunk 1 above)
Comment (NikhilM98) [reply scripts/seed_users.js]: Done:
https://github.com/llaske/sugarizer-server/blob/c020bd58e37b2b48d7071161efdf24d67fec0505/scripts/seed_users.js#L98-L103

File: scripts/seed_users.js (hunk 3 above)
Comment (NikhilM98) [reply scripts/seed_users.js]: Done:
https://github.com/llaske/sugarizer-server/blob/c020bd58e37b2b48d7071161efdf24d67fec0505/scripts/seed_users.js#L79-L84

File: scripts/seed_users.js (hunk 4 above)
Comment (NikhilM98) [reply scripts/seed_users.js]: Done:
https://github.com/llaske/sugarizer-server/blob/c020bd58e37b2b48d7071161efdf24d67fec0505/scripts/seed_users.js#L40-L50
https://github.com/llaske/sugarizer-server/blob/c020bd58e37b2b48d7071161efdf24d67fec0505/scripts/seed_users.js#L147-L151

File: scripts/seed_users.js (hunk 6 above)
Comment (NikhilM98) [reply scripts/seed_users.js]: Done:
https://github.com/llaske/sugarizer-server/blob/c020bd58e37b2b48d7071161efdf24d67fec0505/scripts/seed_users.js#L219-L222

Color is existing classroom is retained.

File: scripts/seed_users.js [hunk 7]
```diff
@@ -0,0 +1,530 @@
...
+        if (validRow) Users.push(new User(validRow.name, validRow.type, validRow.language, validRow.color, validRow.password, validRow.classroom));
+    })
+    .on('end', function() {
+        console.log('Finished processing CSV file');
+        console.log('Starting Sugarizer-Server...');
+        if (Users.length == 0) {
+            console.log('Error: No users to insert');
+            process.exit(-1);
+        }
+        initServer();
```
Comment (llaske) [inline scripts/seed_users.js]: Hmmm. The script should consider that the server is already running. It's strange to be forced to stop the server to launch the script.

Comment (NikhilM98) [reply scripts/seed_users.js]: I wrote the script assuming that the server is not running. I'll remove the `initServer` part.

**Comment by NikhilM98** [comment]
- Reduced verbosity
- Seed assuming the server is already running

Screenshot from 2019-04-13 22-54-47

File: scripts/seed_users.js [hunk 8]
```diff
@@ -31,10 +31,7 @@ var common = require('../dashboard/helper/common');
...
 var xocolors = require('../dashboard/helper/xocolors')();
 
-// Set the node enviornment to test
-process.env.NODE_ENV = 'test';
-
-// Initiate
+// Initialize settings for sugarizer.ini
 var settings;
 
 // Read settings from sugarizer.ini file
```
Comment (llaske) [inline scripts/seed_users.js]: The user should be able to change ini file. In the targeted deployment there is a different ini file for each school. I suggest to read sugarizer.ini except if the ENV is different, like this: https://github.com/llaske/sugarizer-server/blob/dev/config/settings.js#L9

File: scripts/seed_users.js [hunk 9]
```diff
@@ -312,7 +313,6 @@ function getClassroomsNamesFromUsers() {
 
 // Function to generate CSV of users
 function generateCSV() {
-    console.log("Generating CSV...");
     return new Promise(function(resolve, reject) {
         var csvWriter = createCsvWriter({
             path: filename + "_output.csv",
```
Comment (llaske) [inline scripts/seed_users.js]: I understand your idea to both add "output" and let CSV extension, but I'm not a big fan of a so long filename. I suggest to just add .out extension

Comment (NikhilM98) [reply scripts/seed_users.js]: Are you suggesting to add `.out` extension to a `.csv` file? Can you give an example of input and output file names?

Comment (llaske) [reply scripts/seed_users.js]: Ideally the script should have something like a parameter "-o" to specify the output file. In the meantime you could just have an output file named "output.csv".

File: scripts/seed_users.js [hunk 10]
```diff
@@ -0,0 +1,516 @@
...
+    if (!isValidLanguage(user.language)) {
+        user.language = "en";
+    }
+
+    if (!isValidColor(user.color)) {
+        user.color = getRandomColorString();
+    }
+
+    var minPass = (settings && settings.security && parseInt(settings.security.min_password_size)) ? parseInt(settings.security.min_password_size) : 4;
+    if (!user.password || typeof user.password != "string" || user.password.length < minPass) {
```
Comment (llaske) [inline scripts/seed_users.js]: You should validate the password string too. You could use the recently merged function to do that: https://github.com/llaske/sugarizer-server/pull/117/files#diff-d95c18556dca4520ddbc77a44ecc3363

File: scripts/seed_users.js (hunk 9 above)
Comment (NikhilM98) [reply scripts/seed_users.js]: For now, the output will be saved in "output.csv".
https://github.com/llaske/sugarizer-server/blob/50e1614975aaf332b7edcc0efc1f258089e9c228/scripts/seed_users.js#L329-L330

I was thinking about integrating Commander, but I guess it'll be overkill.

File: scripts/seed_users.js (hunk 8 above)
Comment (NikhilM98) [reply scripts/seed_users.js]: Done:
https://github.com/llaske/sugarizer-server/blob/322c84b89518722eee33513759a5b13bdea94839/scripts/seed_users.js#L40-L42

File: scripts/seed_users.js (hunk 10 above)
Comment (NikhilM98) [reply scripts/seed_users.js]: Rebased with upstream/dev and integrated validation:
https://github.com/llaske/sugarizer-server/blob/322c84b89518722eee33513759a5b13bdea94839/scripts/seed_users.js#L34-L36
https://github.com/llaske/sugarizer-server/blob/322c84b89518722eee33513759a5b13bdea94839/scripts/seed_users.js#L140
https://github.com/llaske/sugarizer-server/blob/322c84b89518722eee33513759a5b13bdea94839/scripts/seed_users.js#L160

**Comment by NikhilM98** [comment]
@llaske I've done the requested changes. Please have a look.

**Comment by llaske** [comment]
Good job. I'm waiting for feedback from the deployment.

**Pull request merged**